ext/ScreenShot/ScreenShot.h

```diff
@@ -5,72 +5,186 @@
+			{
+				auto driver = device->getVideoDriver();
+				auto assetManager = device->getAssetManager();
+
+				auto fetchedImageViewParmas = gpuImageView->getCreationParameters();
+				auto gpuImage = fetchedImageViewParmas.image;
+				auto fetchedImageParams = gpuImage->getCreationParameters();
+				auto image = asset::ICPUImage::create(std::move(fetchedImageParams));
+
+				auto trueBufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(fetchedImageParams.extent.width, asset::getTexelOrBlockBytesize(fetchedImageParams.format));
+
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+				asset::ICPUImage::SBufferCopy& region = regions->front();
+
+				region.imageSubresource.mipLevel = 0u;
+				region.imageSubresource.baseArrayLayer = 0u;
+				region.imageSubresource.layerCount = 1u;
+				region.bufferOffset = 0u;
+				region.bufferRowLength = trueBufferRowLength;
+				region.bufferImageHeight = 0u;
+				region.imageOffset = { 0u, 0u, 0u };
+				region.imageExtent = image->getCreationParameters().extent;
+
+				video::IDriverMemoryBacked::SDriverMemoryRequirements memoryRequirements;
+				memoryRequirements.vulkanReqs.alignment = 64u;
+				memoryRequirements.vulkanReqs.memoryTypeBits = 0xffffffffu;
+				memoryRequirements.memoryHeapLocation = video::IDriverMemoryAllocation::ESMT_NOT_DEVICE_LOCAL;
+				memoryRequirements.mappingCapability = video::IDriverMemoryAllocation::EMCF_CAN_MAP_FOR_READ | video::IDriverMemoryAllocation::EMCF_COHERENT | video::IDriverMemoryAllocation::EMCF_CACHED;
+				memoryRequirements.vulkanReqs.size = image->getImageDataSizeInBytes();
+				auto destinationBuffer = driver->createGPUBufferOnDedMem(memoryRequirements);
+
+				auto fence = downloadImageMipLevel(driver, gpuImage.get(), destinationBuffer.get());
+				while (fence->waitCPU(1000ull, fence->canDeferredFlush()) == video::EDFR_TIMEOUT_EXPIRED) {}
 
-	auto buff = core::smart_refctd_ptr<video::IGPUBuffer>(driver->createDownStreamingGPUBufferOnDedMem((asset::getBytesPerPixel(outputFormatOverride)*texSize[0]*texSize[1]).getIntegerApprox()), core::dont_grab); // TODO
-	buff->getBoundMemory()->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ,{0u,buff->getSize()});
+				auto destinationBoundMemory = destinationBuffer->getBoundMemory();
+				destinationBoundMemory->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ, { 0u, memoryRequirements.vulkanReqs.size });
+				auto texelBuffer = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(memoryRequirements.vulkanReqs.size, destinationBoundMemory->getMappedPointer(), core::adopt_memory);
+				
+				auto mapPointerGetterFence = driver->placeFence(true);
+				while (mapPointerGetterFence->waitCPU(1000ull, mapPointerGetterFence->canDeferredFlush()) == video::EDFR_TIMEOUT_EXPIRED) {}
```

Comment:
this is the only fence you need

Reply:
downloadMipMap level function performs `copyImageToBuffer`, doesn't it need a fence?

Reply:
copyImageToBuffer is a GPU-side process, GPU doesn't need to sync with itself (at least not in OpenGL)